### PR TITLE
Add `total_cmp` helper functions to icu_locid types and DataLocale

### PR DESCRIPTION
**Pull request by sffc**
We do not implement `Ord` on `Locale` or `LanguageIdentifier` because there are multiple possible orderings, and we did not want to privilege one ordering over another (see https://github.com/unicode-org/icu4x/issues/1215).

We export a `strict_cmp` function that compares the receiver with a byte array. However, we do not export a function that compares the receiver to another object of the same type. This had led to issues such as https://github.com/unicode-org/icu4x/pull/3588. I also stumbled across this today when trying to make a `BTreeSet` of locales.

In order to serve this use case, I added the following function to the types that needed it:

```rust
    /// Returns an ordering suitable for use in [`BTreeSet`].
    ///
    /// The ordering may or may not be equivalent to string ordering, and it
    /// may or may not be stable across ICU4X releases.
    ///
    /// [`BTreeSet`]: alloc::collections::BTreeSet
    pub fn total_cmp(&self, other: &Self) -> Ordering
```

This gives us flexibility moving forward to tweak the impl or add additional comparison functions producing different orderings.

### CODE REVIEW

File: components/locid/src/extensions/mod.rs [hunk 1]
```diff
@@ -164,6 +167,41 @@ impl Extensions {
             && self.other.is_empty()
     }
 
+    #[allow(clippy::type_complexity)]
+    pub(crate) fn as_tuple(
+        &self,
+    ) -> (
+        (&unicode::Attributes, &unicode::Keywords),
```
Comment (Manishearth) [inline components/locid/src/extensions/mod.rs:174]: ...wow that's a type

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
not 100% convinced this is needed over FFI but it's a weak opinion

File: provider/core/src/request.rs [hunk 2]
```diff
@@ -371,6 +371,24 @@ impl DataLocale {
         self == <&DataLocale>::default()
     }
 
+    /// Returns an ordering suitable for use in [`BTreeSet`].
+    ///
+    /// The ordering may or may not be equivalent to string ordering, and it
+    /// may or may not be stable across ICU4X releases.
+    ///
+    /// [`BTreeSet`]: alloc::collections::BTreeSet
+    pub fn total_cmp(&self, other: &Self) -> Ordering {
```
Comment (robertbastian) [inline provider/core/src/request.rs:380]: remind me why `DataLocale` is not `Ord`?

**Review by robertbastian — APPROVED** [review]

File: provider/core/src/request.rs (hunk 2 above)
Comment (sffc) [reply provider/core/src/request.rs:380]: Same reason `Locale` is not `Ord`, because there are multiple possible orderings and we couldn't agree on which ordering to use by default.

**Review by zbraniecki — APPROVED** [review]
I'd rephrase the comment. Instead of "may or may not be stable", we should say "we do not guarantee stability across versions".

**Review by Manishearth — APPROVED** [review]
